include/graph_contracts.hpp

```diff
@@ -0,0 +1,444 @@
+        } catch (...) {
+            // needed for runtime polymorphism
+            assert(false && "invalid static_cast from 'void* -> WriterType*' in update_reader_internal()");
+            return false;
+        }
+        return true;
+    }
+    friend dyn_port;
+
+    public:
+        constexpr port() = default;
+        constexpr port(std::string_view port_name, std::int16_t priority = 0, std::size_t n_history = 0,
+                       std::size_t min_samples = 0U, std::size_t max_samples = SIZE_MAX) noexcept:
+            _port_name(std::string(port_name)), _priority{priority}, _n_history(n_history),
+            _min_samples(min_samples), _max_samples(max_samples) {
+            static_assert(PortName.empty(), "port name must be exclusively declared via NTTP or constructor parameter");
+        }
+
+        [[nodiscard]] constexpr static port_type_t type() noexcept { return PortType; }
+        [[nodiscard]] constexpr static port_direction_t direction() noexcept { return PortDirection; }
+
+        template<bool enable = !PortName.empty()>
+        [[nodiscard]] static constexpr std::enable_if_t<enable,decltype(PortName)> static_name() noexcept { return PortName; }
+
+    [[nodiscard]] constexpr supported_type pmt_type() const noexcept { return T(); }
+        [[nodiscard]] constexpr std::string_view name() const noexcept {
+            if constexpr (PortName.empty()) {
+                return static_cast<std::string_view>(PortName);
+            } else {
+                return _port_name;
+            }
+        }
+
+        [[nodiscard]] constexpr static bool optional() noexcept { return OPTIONAL; }
+        [[nodiscard]] constexpr std::int16_t priority() const noexcept { return _priority; }
+
+        [[nodiscard]] constexpr static std::size_t available() noexcept { return 0; } //  ↔ maps to Buffer::Buffer[Reader, Writer].available()
+        [[nodiscard]] constexpr std::size_t n_history() const noexcept {
+            if constexpr (N_HISTORY == std::dynamic_extent) {
+                return _n_history;
+            } else {
+                return N_HISTORY;
+            }
+        }
+        [[nodiscard]] constexpr std::size_t min_buffer_size() const noexcept {
+            if constexpr (MIN_SAMPLES == std::dynamic_extent) {
+                return _min_samples;
+            } else {
+                return MIN_SAMPLES;
+            }
+        }
+        [[nodiscard]] constexpr std::size_t max_buffer_size() const noexcept {
+            if constexpr (MAX_SAMPLES == std::dynamic_extent) {
+                return _max_samples;
+            } else {
+                return MAX_SAMPLES;
+            }
+        }
+
+        [[nodiscard]] constexpr connection_result_t resize_buffer(std::size_t min_size) noexcept {
+            if constexpr (IS_INPUT) {
+                static_assert(IS_INPUT, "can only resize output buffers");
+            } else {
+                try {
+                    if constexpr (IS_INPUT) {
```

Comment:
This condition is always false as we're in the `else` branch of `if constexpr (IS_INPUT)`

Reply:
removed dead branch